MaskRCNN_no_batch_convergence/model_frcnn.py

```diff
@@ -420,6 +420,84 @@ def fastrcnn_losses(labels, label_logits, fg_boxes, fg_box_logits):
     return [label_loss, box_loss]
 
 
+@under_name_scope()
+def fastrcnn_predictions_batch(boxes, scores):
+    """
+    Generate final results from predictions of all proposals.
+
+    Args:
+        boxes: n#classx4 floatbox in float32
+        scores: nx#class
+
+    Returns:
+        boxes: Kx4
+        scores: K
+        labels: K
+    """
+    assert boxes.shape[1] == cfg.DATA.NUM_CLASS
+    assert scores.shape[1] == cfg.DATA.NUM_CLASS
+    boxes = tf.transpose(boxes, [1, 0, 2])[1:, :, :]  # #catxnx4
+    scores = tf.transpose(scores[:, 1:], [1, 0])  # #catxn
+
+    def f(X):
+        """
+        prob: n probabilities
+        box: nx4 boxes
+
+        Returns: n boolean, the selection
+        """
+        prob, box = X
+        output_shape = tf.shape(prob, out_type=tf.int64)
+        # filter by score threshold
+        ids = tf.reshape(tf.where(prob > cfg.TEST.RESULT_SCORE_THRESH), [-1])
+        prob = tf.gather(prob, ids)
+        box = tf.gather(box, ids)
+        # NMS within each class
+        #selection = non_max_suppression_custom(
+            #box, prob, cfg.TEST.RESULTS_PER_IM, cfg.TEST.FRCNN_NMS_THRESH)
+        selection = tf.image.non_max_suppression(
+            box, prob, cfg.TEST.RESULTS_PER_IM, cfg.TEST.FRCNN_NMS_THRESH)
+        selection = tf.gather(ids, selection)
+
+        if get_tf_version_tuple() >= (1, 13):
+            sorted_selection = tf.sort(selection, direction='ASCENDING')
+            mask = tf.sparse.SparseTensor(indices=tf.expand_dims(sorted_selection, 1),
+                                          values=tf.ones_like(sorted_selection, dtype=tf.bool),
+                                          dense_shape=output_shape)
+            mask = tf.sparse.to_dense(mask, default_value=False)
+        else:
+            # this function is deprecated by TF
+            sorted_selection = -tf.nn.top_k(-selection, k=tf.size(selection))[0]
+            mask = tf.sparse_to_dense(
+                sparse_indices=sorted_selection,
+                output_shape=output_shape,
+                sparse_values=True,
+                default_value=False)
+        return mask
+
+    # TF bug in version 1.11, 1.12: https://github.com/tensorflow/tensorflow/issues/22750
+    buggy_tf = get_tf_version_tuple() in [(1, 11), (1, 12)]
+    masks = tf.map_fn(f, (scores, boxes), dtype=tf.bool,
+                      parallel_iterations=1 if buggy_tf else 10)     # #cat x N
+    selected_indices = tf.where(masks)  # #selection x 2, each is (cat_id, box_id)
+    scores = tf.boolean_mask(scores, masks)
+
+    # filter again by sorting scores
+    topk_scores, topk_indices = tf.nn.top_k(
```

Comment:
Above says `tf.nn.topk` is deprecated. Should we get rid of it here too? I'm indifferent as long as it works in TF 1.13

Reply:
It works on TF 1.13. We should probably do that later (not as part of this PR) as it has less priority and it will require new convergence runs.